wire-moshi-support/src/main/java/com/squareup/wire/MessageJsonAdapter.java

```diff
@@ -0,0 +1,190 @@
+  }
+
+  @Override public void toJson(JsonWriter out, @Nullable M message) throws IOException {
+    if (message == null) {
+      out.nullValue();
+      return;
+    }
+
+    out.beginObject();
+    for (FieldBinding<M, B> tagBinding : messageAdapter.fieldBindings().values()) {
+      Object value = tagBinding.get(message);
+      if (value == null) {
+        continue;
+      }
+      out.name(tagBinding.name);
+      emitJson(out, value, tagBinding.singleAdapter(), tagBinding.label);
+    }
+    out.endObject();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void emitJson(JsonWriter out, Object value, ProtoAdapter<?> adapter, Label label)
+      throws IOException {
+    if (adapter == ProtoAdapter.UINT64) {
+      if (label.isRepeated()) {
+        List<Long> longs = (List<Long>) value;
+        out.beginArray();
+        for (int i = 0, count = longs.size(); i < count; i++) {
+          emitUint64(longs.get(i), out);
+        }
+        out.endArray();
+      } else {
+        emitUint64((Long) value, out);
+      }
+    } else {
+      moshi.adapter(Object.class).toJson(out, value);
+    }
+  }
+
+  private void emitUint64(Long value, JsonWriter out) throws IOException {
+    if (value < 0) {
+      BigInteger unsigned = POWER_64.add(BigInteger.valueOf(value));
+      out.value(unsigned);
+    } else {
+      out.value(value);
+    }
+  }
+
+  @Override public @Nullable M fromJson(JsonReader in) throws IOException {
+    if (in.peek() == JsonReader.Token.NULL) {
+      in.nextNull();
+      return null;
+    }
+
+    B builder = messageAdapter.newBuilder();
+
+    in.beginObject();
+    while (in.hasNext()) {
+      int index = in.selectName(options);
+      if (index == -1) {
+        in.skipName();
+        in.skipValue();
+        continue;
+      }
+
+      FieldBinding<M, B> fieldBinding = fieldBindings[index];
+      if (fieldBinding == null) {
+        in.skipValue();
+      } else {
+        Object value = parseValue(fieldBinding, in);
+        fieldBinding.set(builder, value);
+      }
+    }
+
+    in.endObject();
+    return builder.build();
+  }
+
+  private Object parseValue(FieldBinding<?, ?> fieldBinding, JsonReader in) throws IOException {
+    ProtoAdapter<?> protoAdapter = fieldBinding.singleAdapter();
+
+    if (protoAdapter == ProtoAdapter.UINT64) {
+      if (fieldBinding.label.isRepeated()) {
+        List<Long> longs = new ArrayList<>();
+        in.beginArray();
+        while (in.hasNext()) {
+          longs.add(consumeUint64(in));
+        }
+        in.endArray();
+
+        return longs;
+      } else {
+        return consumeUint64(in);
+      }
+    }
+
+    if (fieldBinding.label.isRepeated()) {
+      if (in.peek() == JsonReader.Token.NULL) {
+        in.nextNull();
+        return Collections.emptyList();
+      }
+
+      Class<?> itemType = protoAdapter.javaType;
+
+      return moshi.adapter(Types.newParameterizedType(List.class, itemType)).fromJson(in);
```

Comment:
Lots of room for future optimization like caching these adapters so we don't have to allocate each time we want to use it

Reply:
Isn't moshi doing the caching here?
https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/Moshi.java#L94-L148

Reply:
Yes but it requires allocating the parameterized type. We should be able to maintain an array of adapters which are parallel to the field bindings so that it's O(1) lookups. Plus it pushes validation to creation time instead of serialization time.

Reply:
Ok, I can do that in a follow up.
I believe I'll have to create a `3 * size` array of adapters so each field has its adapter, its list adapter, and the map one, does that sound correct?

Reply:
Hmm I thought we would only need one per field. I don't think it's at all urgent, we can play around in a follow-up.

Reply:
fixed